pulsar-transaction/coordinator/src/test/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriterTest.java

```diff
@@ -613,9 +610,371 @@ public ByteBuf serialize(ArrayList<Integer> dataArray) {
+        SumStrDataSerializer dataSerializer = new SumStrDataSerializer();
+        int writeCount = 100;
+        int expectedTotalBytesSize = writeCount * dataSerializer.getSizePerData();
+        var callbackWithCounter = createCallBackWithCounter();
+        // Create TxnLogBufferedWriter.
+        var txnLogBufferedWriterContext =
+                createTxnBufferedWriterContextWithMetrics(dataSerializer, Integer.MAX_VALUE,
+                        Integer.MAX_VALUE, 1);
+        var txnLogBufferedWriter = txnLogBufferedWriterContext.txnLogBufferedWriter;
+        // Add some data.
+        for (int i = 0; i < writeCount; i++){
+            txnLogBufferedWriter.asyncAddData(1, callbackWithCounter.callback, "");
+            Thread.sleep(1);
+        }
+        // Wait for all write finish.
+        Awaitility.await().atMost(5, TimeUnit.SECONDS).until(
+                () -> callbackWithCounter.finishCounter.get() + callbackWithCounter.failureCounter.get() == writeCount
+        );
+        int actualBatchFlushCount = txnLogBufferedWriterContext.mockedManagedLedger.writeCounter.get();
+        assertEquals(callbackWithCounter.failureCounter.get(), 0);
+        verifyTheCounterMetrics(0,0, actualBatchFlushCount,0);
+        verifyTheHistogramMetrics(actualBatchFlushCount, writeCount, expectedTotalBytesSize);
+        // cleanup.
+        releaseTxnLogBufferedWriterContext(txnLogBufferedWriterContext);
+        // after close, verify the metrics change to 0.
+        verifyTheCounterMetrics(0,0,0,0);
+        verifyTheHistogramMetrics(0,0,0);
+    }
+
+    @Test
+    public void testMetricsStatsThatTriggeredByLargeSingleData() throws Exception {
+        // Use TwoLenSumDataSerializer for: write a little data once, then write a large data once.
+        int bytesSizePerRecordWhichInBatch = 4;
+        int batchedWriteMaxSize = 1024;
+        TwoLenSumDataSerializer dataSerializer =
+                new TwoLenSumDataSerializer(bytesSizePerRecordWhichInBatch, batchedWriteMaxSize);
+        int writeCount = 100;
+        int singleLargeDataRequestCount = writeCount / 2;
+        int expectedBatchFlushTriggeredByLargeData = singleLargeDataRequestCount;
+        int expectedTotalBytesSize = expectedBatchFlushTriggeredByLargeData * bytesSizePerRecordWhichInBatch;
+        var callbackWithCounter = createCallBackWithCounter();
+        // Create TxnLogBufferedWriter.
+        var txnLogBufferedWriterContext = createTxnBufferedWriterContextWithMetrics(
+                dataSerializer, Integer.MAX_VALUE, batchedWriteMaxSize, Integer.MAX_VALUE);
+        var txnLogBufferedWriter = txnLogBufferedWriterContext.txnLogBufferedWriter;
+        // Add some data.
+        for (int i = 0; i < writeCount; i++){
+            txnLogBufferedWriter.asyncAddData(1, callbackWithCounter.callback, i);
+        }
+        // Wait for all data write finish.
+        Awaitility.await().atMost(2, TimeUnit.SECONDS).until(
```

Comment:
Where are you checking that you wrote to managed ledger writeCount times?

I think this test is hard to understand.

I would do:
```
    createBufferedWriter(maxBatchSize = batchedWriteMaxSize, maxRecords = MAX_INT, maxDelay = MAX_INT)
    writeToBuffer(randomRecord(size = batchedWriteMaxSize/10)
    writeToBuffer(randomRecord(size = batchedWriteMaxSize/10)
    writeToBuffer(randomRecord(size = maxBatchSize)
    waitForAllWriteToCompleteSuccessfully()
    assertThat(actualNumOfFlushes, 1)
    assertThat(actualWritesToManagedLedger,2)
```

Reply:
It seems to make no difference. In this unit test, we use `TwoLenSumDataSerializer` to control the size of each write: [4, batchedWriteMaxSize, batchedWriteMaxSize, 4, batchedWriteMaxSize....]

Reply:
You need to check :
>Where are you checking that you wrote to managed ledger writeCount times?

Reply:
> Where are you checking that you wrote to managed ledger writeCount times?

Already add this check